designs/memory-management-design.md

```diff
@@ -0,0 +1,275 @@
+    return r
+
+x: Rat = None
+x = someFunc()
+```
+**Expected:**
+Let `o` be the object referred to by variable `x`
+```
+assert number of references of o is 1
+assert size of o(in bytes) is 4
+assert type of the fields in o is [value]
+```
+
+### 6. Access is not assignment
+
+**Case:**
+```
+class Rat(object):
+    y: int = 0
+    def __init__(self: Rat):
+		self.y = 1
+
+x: Rat = None
+x = Rat()
+x.y
+print(x.y)
+```
+**Expected:**
+Let `o` be the object referred to by variable `x`
+```
+assert number of references of o is 1
+assert size of o(in bytes) is 4
+assert type of the fields in o is [value]
+```
+
+### 7: Objects as fields
+
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+    def add(l: Link) -> Link:
+        m: Link = None
+        m = Link()
+        l.next = m
+        return m
+
+x: Link = None
+y: Link = None
+y = x.add()
+```
+**Expected:**
+
+Let `o` be the object referred to by variable `x` <br>
+Let `p` be the object referred to by variable `y`
+```
+assert number of references of o is 1
+assert type of fields in o is [value, pointer]
+assert number of references of p is 2
+assert type of fields in p is [value, pointer]
+```
+
+### 8. Anonymous object deletion
+
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+    def add(l: Link) -> Link:
+        l.next = Link()
+        return l.next
+
+x: Link = None
+x.add()
+x = None
+```
+**Expected:**
+```
+assert number of references of any object is 0
+```
+
+### 9. Simple linked cycle
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+
+x: Link = None
+y: Link = None
+x = Link()
+y = Link()
+x.next = y
+y.next = x
+```
+**Expected:**
+
+Let `o` be the object referred to by variable `x` <br>
+Let `p` be the object referred to by variable `y`
+```
+assert number of references of o is 1
+assert type of fields in o is [value, pointer]
+assert number of references of p is 1
+assert type of fields in p is [value, pointer]
+```
+
+### 10. Simple deletion in cycle
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+
+x: Link = None
+y: Link = None
+x = Link()
+y = Link()
+x.next = y
+y.next = x
+
+x = None
+```
+**Expected:**
+
+Let `o` be the object referred to by variable `x` <br>
+Let `p` be the object referred to by variable `y`
+```
+assert number of references of o is 0
+assert type of fields in o is [value, pointer]
+assert number of references of p is 1
+assert type of fields in p is [value, pointer]
+```
+
+## Changes to IR
+
+- ### Addition of `ref` value
+```
+export type Value<A> =
+...
+| { a?: A, tag: "ref"}
+```
+Currently there is no way of differentiating between pointers and values stored in an object. As seen in test case *8: Anonymous object deletion*, we need to follow references stored in an object recursively so that we can update their respective reference counts. This is not possible without knowing which field represents a reference/pointer.
```

Comment:
I'm confused about this. There's no runtime information on `ref`, so it's not clear what this will compile to/be stored as. Can't you find out which fields are references/pointers by looking up the class and using type information? If it's bool/int then it's not a pointer, if it's class then it is? Statically I think this isn't a problem.